game-core/src/main/java/games/strategy/engine/framework/ServerGame.java

```diff
@@ -442,22 +442,23 @@ private void runStep(final boolean stepIsRestoredFromSavedGame) {
     if (gameData.getSequence().next()) {
       gameData.getHistory().getHistoryWriter().startNextRound(gameData.getSequence().getRound());
       saveGame(gameData.getSequence().getRound() % 2 == 0
-          ? AutoSaveFileUtils.getEvenRoundAutoSaveFile(headless)
-          : AutoSaveFileUtils.getOddRoundAutoSaveFile(headless));
+          ? launchAction.getAutoSaveFileUtils().getEvenRoundAutoSaveFile()
+          : launchAction.getAutoSaveFileUtils().getOddRoundAutoSaveFile());
     }
     if (autoSaveThisDelegate && !currentStep.getName().endsWith("Move")) {
-      autoSaveAfter(currentDelegate, headless);
+      autoSaveAfter(currentDelegate);
     }
   }
 
-  private void autoSaveAfter(final String stepName, final boolean headless) {
-    saveGame(AutoSaveFileUtils.getAfterStepAutoSaveFile(stepName, headless));
+  private void autoSaveAfter(final String stepName) {
+    final var saveUtils = launchAction.getAutoSaveFileUtils();
+    saveGame(saveUtils.getAfterStepAutoSaveFile(saveUtils.getAutoSaveStepName(stepName)));
   }
 
-  private void autoSaveAfter(final IDelegate delegate, final boolean headless) {
+  private void autoSaveAfter(final IDelegate delegate) {
     final String typeName = delegate.getClass().getTypeName();
     final String stepName = typeName.substring(typeName.lastIndexOf('.') + 1).replaceFirst("Delegate$", "");
-    saveGame(AutoSaveFileUtils.getAfterStepAutoSaveFile(stepName, headless));
+    saveGame(launchAction.getAutoSaveFileUtils().getAfterStepAutoSaveFile(stepName));
```

Comment:
See a law of demeter violation here, ie: `a.b().c()`, which can be converted to `a.d()`, ie: `launchAction.getAfterStepAutoSaveFile(stepName)`. I think perhaps we might actually want two objects, a `launchAction` and a `saveGameActions`. In that case we could call something like `saveFileUtils.getAfterStepAutoSaveFile(stepName)`

Then also, `ServerGame` knows far too much about type and step names to formulate a file name. Perhaps these three lines from `autoSaveAfter` should be more simply: `saveGame(saveFileUtils.generateAfterStepFileName(delegate))`